### PR TITLE
Add snapshot and rollup validations

### PR DESCRIPTION
**Pull request by animesh2049**
This PR introduces two valdation functions, `verifySnapshot` and `verifyPostingSplits` for snapshots and rollup respectively. 
`verifySnapshot` is called after snapshot, it iterates over all the keys in badger. For every split data key, it checks if the part keys are also present.
`verifyPostingSplits` is called after rollup. It checks if every part key is present in the  key value slice returned by rollup. 

**Validations are run only when dgraph is compiled with `debug` build flag**

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6397)

### CODE REVIEW

**Review by martinmr — COMMENTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6397)* status: all files reviewed, 3 unresolved discussions (waiting on @animesh2049)

___
*[worker/snapshot.go, line 123 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MGP-LeNDnSXEhD76wyB:-MGP-LeNDnSXEhD76wyC:bbxp0a2) ([raw file](https://github.com/dgraph-io/dgraph/blob/1ebd5a35031a22a1a1927da1ea305629e8ccdbb1/worker/snapshot.go#L123)):*
> ```Go
> 	iOpt := badger.DefaultIteratorOptions
> 	iOpt.AllVersions = true
> 	txn := pstore.NewTransactionAt(math.MaxUint64, false)
> ```

use the same timestamp the snapshot is using.
___
*[worker/snapshot.go, line 124 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MGP-D6y1tgFzs-kfWy8:-MGP-D6y1tgFzs-kfWy9:b-6x9qph) ([raw file](https://github.com/dgraph-io/dgraph/blob/1ebd5a35031a22a1a1927da1ea305629e8ccdbb1/worker/snapshot.go#L124)):*
> ```Go
> 	iOpt.AllVersions = true
> 	txn := pstore.NewTransactionAt(math.MaxUint64, false)
> 	it := txn.NewIterator(iOpt)
> ```

you can use the stream framework for this instead of a single iterator.
___
*[worker/snapshot.go, line 142 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MGP-P0SA8g5f8Win5wH:-MGP-P0SA8g5f8Win5wI:b-fxy2qk) ([raw file](https://github.com/dgraph-io/dgraph/blob/1ebd5a35031a22a1a1927da1ea305629e8ccdbb1/worker/snapshot.go#L142)):*
> ```Go
> 			sKey, kErr := x.SplitKey(k, uid)
> 			x.Check(kErr)
> 			newTxn := pstore.NewTransactionAt(math.MaxUint64, false)
> ```

use the same timestamp as the version of the key. The main list and all parts are written at the same timestamp.

**Review by animesh2049 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6397)* status: 1 of 2 files reviewed, 3 unresolved discussions (waiting on @martinmr)

___
*[worker/snapshot.go, line 123 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MGP-LeNDnSXEhD76wyB:-MGbf8Lw6Psyu5V-zmY0:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1ebd5a35031a22a1a1927da1ea305629e8ccdbb1/worker/snapshot.go#L123)):*
<details><summary><i>Previously, martinmr (Martin Martinez Rivera) wrote…</i></summary><blockquote>

use the same timestamp the snapshot is using.
</blockquote></details>

Done.
___
*[worker/snapshot.go, line 124 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MGP-D6y1tgFzs-kfWy8:-MGbf9kL3Vx-qshmvxhW:b3acmvs) ([raw file](https://github.com/dgraph-io/dgraph/blob/1ebd5a35031a22a1a1927da1ea305629e8ccdbb1/worker/snapshot.go#L124)):*
<details><summary><i>Previously, martinmr (Martin Martinez Rivera) wrote…</i></summary><blockquote>

you can use the stream framework for this instead of a single iterator.
</blockquote></details>

Leaving it for now. Need to check the failing tests.
___
*[worker/snapshot.go, line 142 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MGP-P0SA8g5f8Win5wH:-MGbf9MqAWPhSHgs4Nr1:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1ebd5a35031a22a1a1927da1ea305629e8ccdbb1/worker/snapshot.go#L142)):*
<details><summary><i>Previously, martinmr (Martin Martinez Rivera) wrote…</i></summary><blockquote>

use the same timestamp as the version of the key. The main list and all parts are written at the same timestamp.
</blockquote></details>

Done.

File: x/debug_off.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// +build !debug
+
```
Comment (jarifibrahim) [inline x/debug_off.go:2]: missing license

File: x/debug_off.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	bpb "github.com/dgraph-io/badger/v2/pb"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/dgraph-io/dgraph/protos/pb"
+)
+
+// VerifySnapshot works in debug mode. Check out the comment in debug_on.go
+func VerifySnapshot(pstore *badger.DB, readTs uint64) {
+
```
Comment (jarifibrahim) [inline x/debug_off.go:14]: remove newline

File: x/debug_off.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+// VerifyPostingSplits works in debug mode. Check out the comment in debug_on.go
+func VerifyPostingSplits(kvs []*bpb.KV, plist *pb.PostingList,
+	parts map[uint64]*pb.PostingList, baseKey []byte) {
+
```
Comment (jarifibrahim) [inline x/debug_off.go:20]: remove newline

File: x/debug_on.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
+// +build debug
+
```
Comment (jarifibrahim) [inline x/debug_on.go:2]: missing license

File: x/debug_on.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	it := txn.NewIterator(iOpt)
+	for it.Rewind(); it.Valid(); it.Next() {
+		i := it.Item()
+		k := i.Key()
+		parsedKey, kErr := Parse(k)
+		Check(kErr)
+		if !parsedKey.IsData() {
+			continue
+		}
+		v, vErr := i.ValueCopy(nil)
```
Comment (jarifibrahim) [inline x/debug_on.go:30]: You can avoid copying and use the `i.Value(func(v []byte){})`

File: x/debug_on.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+		Check(vErr)
+		plist := &pb.PostingList{}
+		err := plist.Unmarshal(v)
```
Comment (jarifibrahim) [inline x/debug_on.go:33]: check(plist.unmarshal)

File: x/debug_on.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+		Check(err)
+		if len(plist.Splits) == 0 {
+			continue
+		}
+		if plist.Splits[0] != uint64(1) {
+			log.Panic("First split UID is not 1")
+		}
+		for _, uid := range plist.Splits {
+			sKey, kErr := SplitKey(k, uid)
+			Check(kErr)
```
Comment (jarifibrahim) [inline x/debug_on.go:43]: key+version. Anything that might help us debug this.

File: x/debug_on.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		v, vErr := i.ValueCopy(nil)
+		Check(vErr)
+		plist := &pb.PostingList{}
+		err := plist.Unmarshal(v)
+		Check(err)
+		if len(plist.Splits) == 0 {
+			continue
+		}
+		if plist.Splits[0] != uint64(1) {
+			log.Panic("First split UID is not 1")
```
Comment (jarifibrahim) [inline x/debug_on.go:39]: Dump key and version

File: x/debug_on.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+		parsedKey, kErr := Parse(k)
+		Check(kErr)
+		if !parsedKey.IsData() {
+			continue
+		}
+		v, vErr := i.ValueCopy(nil)
+		Check(vErr)
+		plist := &pb.PostingList{}
+		err := plist.Unmarshal(v)
+		Check(err)
```
Comment (jarifibrahim) [inline x/debug_on.go:34]: Dump key and version while crashing.

File: x/debug_on.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+	for it.Rewind(); it.Valid(); it.Next() {
+		i := it.Item()
+		k := i.Key()
+		parsedKey, kErr := Parse(k)
+		Check(kErr)
+		if !parsedKey.IsData() {
+			continue
+		}
+		v, vErr := i.ValueCopy(nil)
+		Check(vErr)
```
Comment (jarifibrahim) [inline x/debug_on.go:31]: Dump the key and its version while crashing.

File: x/debug_on.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+func VerifySnapshot(pstore *badger.DB, readTs uint64) {
+	iOpt := badger.DefaultIteratorOptions
+	iOpt.AllVersions = true
+	txn := pstore.NewTransactionAt(readTs, false)
+	it := txn.NewIterator(iOpt)
+	for it.Rewind(); it.Valid(); it.Next() {
+		i := it.Item()
+		k := i.Key()
+		parsedKey, kErr := Parse(k)
+		Check(kErr)
```
Comment (jarifibrahim) [inline x/debug_on.go:26]: Dump the byte slice along with the error.

File: x/debug_on.go [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+					sKey, "\nbaseKey: ", parsedKey, "\nSplits: ", plist.Splits,
+				)
+			}
+		}
+	}
+}
+
+func VerifyPostingSplits(kvs []*bpb.KV, plist *pb.PostingList,
+	parts map[uint64]*pb.PostingList, baseKey []byte) {
+
```
Comment (jarifibrahim) [inline x/debug_on.go:57]: remove new line.

File: x/debug_on.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+			if dbErr != nil {
+				log.Panic("Snapshot verification failed: Unable to find splitKey: ",
+					sKey, "\nbaseKey: ", parsedKey, "\nSplits: ", plist.Splits,
+				)
+			}
+		}
+	}
+}
+
+func VerifyPostingSplits(kvs []*bpb.KV, plist *pb.PostingList,
```
Comment (jarifibrahim) [inline x/debug_on.go:55]: Needs a description.

**Review by jarifibrahim — COMMENTED** [review]
Got some comments. This PR needs a proper title and description.

**Review by animesh2049 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6397)* status: 0 of 4 files reviewed, 16 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[x/debug_off.go, line 2 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy--r2-2:-MHGVbrs9I3kRcahHlqs:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_off.go#L2)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

missing license
</blockquote></details>

Done.
___
*[x/debug_off.go, line 14 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy--r2-14:-MHGVe4k1YqMwik4h-Er:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_off.go#L14)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

remove newline
</blockquote></details>

Done.
___
*[x/debug_off.go, line 20 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy--r2-20:-MHGVeNaAEvL3TsHKph8:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_off.go#L20)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

remove newline
</blockquote></details>

Done.
___
*[x/debug_on.go, line 2 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-2:-MHGVmNACpGUqVb5UCCn:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L2)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

missing license
</blockquote></details>

Done.
___
*[x/debug_on.go, line 26 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-26:-MHGVpGH3acxpiyTH6uU:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L26)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Dump the byte slice along with the error.
</blockquote></details>

Done.
___
*[x/debug_on.go, line 30 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-30:-MHGUA-cDLldaeDQPD_-:b-u1lal) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L30)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

You can avoid copying and use the `i.Value(func(v []byte){})`
</blockquote></details>

We will have to put the entire code inside the function then. That will be one more level of indentation.
___
*[x/debug_on.go, line 31 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-31:-MHGWBsDEbIvovP5rR6C:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L31)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Dump the key and its version while crashing.
</blockquote></details>

Done.
___
*[x/debug_on.go, line 33 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-33:-MHGWCMi3EXDW7lbFgiy:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L33)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

check(plist.unmarshal)
</blockquote></details>

Done.
___
*[x/debug_on.go, line 34 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-34:-MHGWFm7A4xElYBOlXET:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L34)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Dump key and version while crashing.
</blockquote></details>

Done.
___
*[x/debug_on.go, line 39 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-39:-MHGWGh49PwLVKeFsb55:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L39)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Dump key and version
</blockquote></details>

Done.
___
*[x/debug_on.go, line 43 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-43:-MHGWyimBYotdY2Qtb9f:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L43)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

key+version. Anything that might help us debug this.
</blockquote></details>

Done.
___
*[x/debug_on.go, line 55 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-55:-MHGWzw_5fd2Ot0E_B3w:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L55)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Needs a description.
</blockquote></details>

Done.
___
*[x/debug_on.go, line 57 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-57:-MHGS_KX-WahsvOS0aXp:bqrvkhl) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L57)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

remove new line.
</blockquote></details>

function declaration is in 2 lines should there be a newline after declaration ?

**Review by jarifibrahim — APPROVED** [review]
:lgtm: Get final approval from @martinmr  

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-:-MHKt3nr72daZgEbM7cA:b-metri4)* status: 0 of 4 files reviewed, 8 unresolved discussions (waiting on @animesh2049, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[x/debug_on.go, line 30 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-30:-MHKs0Db7hfVQhQh9REp:b-mypumk) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L30)):*
<details><summary><i>Previously, animesh2049 (Animesh Chandra Pathak) wrote…</i></summary><blockquote>

We will have to put the entire code inside the function then. That will be one more level of indentation.
</blockquote></details>

this will create many small allocations and will cause stress on gc . Indentation is okay.
___
*[x/debug_on.go, line 57 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-57:-MHKsIn_EWzaUfnc2hCX:bsbl8y8) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L57)):*
<details><summary><i>Previously, animesh2049 (Animesh Chandra Pathak) wrote…</i></summary><blockquote>

function declaration is in 2 lines should there be a newline after declaration ?
</blockquote></details>

we don't need the new line
___
*[x/debug_on.go, line 2 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHKsuQtCG_R39B7gC3Y:-MHKsuQtCG_R39B7gC3Z:bdxpaw2) ([raw file](https://github.com/dgraph-io/dgraph/blob/a8a06902f9c8e46e0701ea0b053507fe94f15cf5/x/debug_on.go#L2)):*
> ```Go
> // +build debug
> ```

Call this file `debug.go` and the other one `nodebug.go`
___
*[x/debug_on.go, line 38 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHKsQYI18K6KWs1Tx15:-MHKsQYI18K6KWs1Tx16:bcdeiwr) ([raw file](https://github.com/dgraph-io/dgraph/blob/a8a06902f9c8e46e0701ea0b053507fe94f15cf5/x/debug_on.go#L38)):*
> ```Go
> 	txn := pstore.NewTransactionAt(readTs, false)
> 	it := txn.NewIterator(iOpt)
> 	for it.Rewind(); it.Valid(); it.Next() {
> ```

@martinmr suggested using stream framework. Is there a reason for not using it?
___
*[x/debug_on.go, line 72 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHKsG9N9AQ21wQ20zUv:-MHKsG9N9AQ21wQ20zUw:b-umah8b) ([raw file](https://github.com/dgraph-io/dgraph/blob/a8a06902f9c8e46e0701ea0b053507fe94f15cf5/x/debug_on.go#L72)):*
> ```Go
> 
> // VerifyPostingSplits checks if all the keys from parts are
> // present in kvs. parts is a map of split keys -> postinglist
> ```

nit: parts => Parts and missing period.

**Review by animesh2049 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6397)* status: 0 of 4 files reviewed, 7 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[x/debug_on.go, line 30 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-30:-MHL_3ZK1LOQbsdwKiLd:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L30)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

this will create many small allocations and will cause stress on gc . Indentation is okay.
</blockquote></details>

Done.
___
*[x/debug_on.go, line 57 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHFrXSmX_tjN-YCkQy0-r2-57:-MHL_4V_Bx5kFH2zddB1:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/faa7cf801784f37af239f70725a00c8d0402d4fc/x/debug_on.go#L57)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

we don't need the new line
</blockquote></details>

Done.
___
*[x/debug_on.go, line 2 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHKsuQtCG_R39B7gC3Y:-MHKxGad49ONjx7qXt9n:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/a8a06902f9c8e46e0701ea0b053507fe94f15cf5/x/debug_on.go#L2)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Call this file `debug.go` and the other one `nodebug.go`
</blockquote></details>

Done.
___
*[x/debug_on.go, line 38 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6397#-MHKsQYI18K6KWs1Tx15:-MHKvPqX7M4Q-qaCs9Vw:bqdlzi9) ([raw file](https://github.com/dgraph-io/dgraph/blob/a8a06902f9c8e46e0701ea0b053507fe94f15cf5/x/debug_on.go#L38)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

@martinmr suggested using stream framework. Is there a reason for not using it?
</blockquote></details>

Not really. Earlier there was no plan of check this into our code and I didn't had idea about how stream framework works.

**Pull request merged**